pkg/portproxy/server.go

```diff
@@ -0,0 +1,147 @@
+	quit     chan struct{}
+	// map of port number as a key to associated listener
+	activeListeners map[int]net.Listener
+	mutex           sync.Mutex
+	wg              sync.WaitGroup
+}
+
+func NewPortProxy(address string) (*PortProxy, error) {
+	socket, err := net.Listen("unix", address)
+
+	portProxy := &PortProxy{
+		address:         address,
+		listener:        socket,
+		quit:            make(chan struct{}),
+		activeListeners: make(map[int]net.Listener),
+	}
+	return portProxy, err
+}
+
+func (p *PortProxy) Listen() error {
+	logrus.Infof("Proxy server started listening on %s, forwarding to %s", p.address, bridgeIPAddr)
+	for {
+		conn, err := p.listener.Accept()
+		if err != nil {
+			select {
+			case <-p.quit:
+				return nil
+			default:
+				return fmt.Errorf("published port server connection accept error: %w", err)
+			}
+		} else {
+			go p.handleEvent(conn)
+		}
+	}
+}
+
+func (p *PortProxy) handleEvent(conn net.Conn) {
+	defer conn.Close()
+
+	var pm types.PortMapping
+	err := json.NewDecoder(conn).Decode(&pm)
+	if err != nil {
+		logrus.Errorf("port server decoding received payload error: %s", err)
+		return
+	}
+	p.execListener(pm)
+}
+
+func (p *PortProxy) execListener(pm types.PortMapping) {
+	for _, portBindings := range pm.Ports {
+		for _, portBinding := range portBindings {
+			logrus.Debugf("received the following port: [%s] from portMapping: %+v", portBinding.HostPort, pm)
+			port, err := nat.ParsePort(portBinding.HostPort)
+			if err != nil {
+				logrus.Errorf("parsing port error: %s", err)
+				continue
+			}
+			p.mutex.Lock()
+			if pm.Remove {
+				if listener, exist := p.activeListeners[port]; exist {
+					if err := listener.Close(); err != nil {
+						logrus.Errorf("error closing listener for port [%s]: %s", portBinding.HostPort, err)
+					}
+				}
+				p.mutex.Unlock()
+				continue
+			}
+			p.mutex.Unlock()
+			addr := net.JoinHostPort("localhost", portBinding.HostPort)
+			l, err := net.Listen("tcp", addr)
+			if err != nil {
+				logrus.Errorf("failed creating listener for published port [%s]: %s", portBinding.HostPort, err)
+				continue
+			}
+			p.mutex.Lock()
+			p.activeListeners[port] = l
+			p.mutex.Unlock()
+			go p.acceptTraffic(l, portBinding.HostPort)
+		}
+	}
+}
+
+func (p *PortProxy) acceptTraffic(listener net.Listener, port string) {
+	forwardAddr := net.JoinHostPort(bridgeIPAddr, port)
+	for {
+		conn, err := listener.Accept()
+		if err != nil {
+			// Check if the error is due to listener being closed
+			if errors.Is(err, net.ErrClosed) {
+				break
+			}
+			logrus.Errorf("port proxy listener failed to accept: %s", err)
+			continue
+		}
+		logrus.Debugf("port proxy accepted connection from %s", conn.RemoteAddr())
+		p.wg.Add(1)
+
+		go func(conn net.Conn) {
+			defer p.wg.Done()
+			defer conn.Close()
+			utils.Pipe(conn, forwardAddr)
+		}(conn)
+	}
+}
+
+func (p *PortProxy) Wait() {
+	p.wg.Wait()
+}
+
+func (p *PortProxy) Close() error {
+	close(p.quit)
+	return p.listener.Close()
```

Comment:
Shouldn't we close the listener before closing the channel (so that we don't have an instant where we can get a new connection but the channel is already closed)?